crates/reporting/tests/test_reporting.rs

```diff
@@ -10487,4 +10518,38 @@ All branches in an `if` must have the same type!
     Note: `Decoding` cannot be generated for functions.
     "###
     );
+
+    test_report!(
+        record_with_optional_field_types_cannot_derive_decoding,
+        indoc!(
+            r#"
+             app "test" imports [Decode.{Decoder, DecoderFormatting, decoder}] provides [main] to "./platform"
+
+             main =
+                 myDecoder : Decoder {x : Str, y ? Str} fmt | fmt has DecoderFormatting
+                 myDecoder = decoder
+
+                 myDecoder
+             "#
+        ),
+        @r###"
+     ── TYPE MISMATCH ───────────────────────────────────────── /code/proj/Main.roc ─
+
+     This expression has a type that does not implement the abilities it's expected to:
```

Comment:
Not a blocker, but I'd shorten this to:

> This expression can't be used here because it's missing a required ability:

Also this would be a nice place for one of those "more specific alternatives to the word `expression`" (I forget what they're called) - e.g. "This `decoder` record" instead of "This expression"